dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/EventVisitor.java

```diff
@@ -0,0 +1,78 @@
+package com.dat3m.dartagnan.program.event.visitors;
+
+import com.dat3m.dartagnan.program.event.arch.tso.*;
+import com.dat3m.dartagnan.program.event.arch.aarch64.*;
+import com.dat3m.dartagnan.program.event.core.*;
+import com.dat3m.dartagnan.program.event.core.rmw.*;
+import com.dat3m.dartagnan.program.event.lang.catomic.*;
+import com.dat3m.dartagnan.program.event.lang.linux.*;
+import com.dat3m.dartagnan.program.event.lang.linux.cond.*;
+import com.dat3m.dartagnan.program.event.lang.pthread.*;
+import com.dat3m.dartagnan.program.event.lang.svcomp.*;
+
+public interface EventVisitor<T> {
+    
+	// Basic events
+	default T visit(Assume e) { return visit((Event)e); };
+	default T visit(Cmp e) { return visit((Skip)e); };
+	default T visit(CondJump e) { return visit((Event)e); };
+	default T visit(Event e) { return visit((Event)e); };
```

Comment:
Endless recursion incoming :).

I'm thinking that it might be better to give the different `visit` methods also different names.
While it is somewhat easier to not do so, it is always easy to accidentally call the wrong methods.

Reply:
Instead of null, I throw an exception. Each concrete class implementing a visitor should specify the behaviour for `visitEvent()` since this is the default method call for any method that is not override.

I also added the event class name to each method to avoid problems (this is also what the automatically generated visitor for antlr do).

Reply:
I don't think visitors have to specify a behavior for the default case. Despite the interface, visitors do not have to have a return value and in fact, you probably wouldn't use return values for quite a few cases like e.g. implementing a ExprSimplifierVisitor which only visits events like `MemEvent` and `Local`.
And for these types of visitors, visiting any type of event they do not care about is basically a NOOP.
If the interface forces an exception there, you would need to overwrite the default behaviour with empty methods which is kinda ugly. Also, if you want this method to be overwritten in the first place, then you could simply avoid making it default.

Reply:
By "visitors have to specify a behavior for the default case" I meant "for visitEvent()". All default implementations in the interface end up pointing to that method. Therefore, if a concrete visitor just overrides a couple of methods (which will probably be the case for most use-cases), for the remaining ones `visitorEvent()` will be called and thus that one must have a proper implementation. But the implementation is use-case specific (e.g., for compilation is just returns a singleton list with the input). You proposed to use `null` in this case, but that is dangerous because unless the method using the return of the visitor checks for not null, this would be unnoticed. That is why I think throwing an exception (only) for `visitEvent()` in the interface to force the programmer to implement **that** method for each visitor is the way to go.

Reply:
No, if you want the implementor to implement `visitEvent` you simply make it non-default. Then it is impossible to forget it.
But you argue about return types and my argument is that not all visitors have return types. Those that do will have to overwrite `visitEvent, those that do not will not want to do so. And I think there are many cases where the visitor does not use return types, possibly more than there are that use return types.

Reply:
I think that setting the default implementation of `visitEvent()` to `null` is just too dangerous. I would remove the default implementation of it, thus forcing the developer to provide an implementation, and if in that particular case a NOOP is needed, `null` can be provided as an implementation.

Reply:
Removing the default implementation is ok, it is better than the default exception. Actually, throwing an exception as default behaviour in any interface or class is always stupid because the methods can be made non-default/abstract instead (I think we have this case in some abstract RMW base classes).